arg_spec - rework _check_arguments()

samdoran opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Moving towards controller side arg spec validation. Part of role argument spec validation for 2.11.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/module_utils/basic.py`
`lib/ansible/module_utils/common/parameters.py`

samdoran committed:
Move _syslog_facitily to __init__

No good reason it should not be set for each object

samdoran committed:
Move internal property setting to private method

samdoran committed:
Create check_arguments() function

samdoran committed:
Remove unused import

samdoran committed:
Rename function to better match its behavior

Change the behavior to return a set, either empty or populated, with unsupported keys.
Accept legal_inputs as optional which will not required calling handle_aliases before calling
get_unsupported_parameters().

samdoran committed:
Add changelog

Shrews commented on lib/ansible/module_utils/common/parameters.py:
```diff
@@ -195,3 +195,29 @@ def handle_aliases(argument_spec, params, alias_warnings=None):
+
+
+def get_unsupported_parameters(argument_spec, module_parameters, legal_inputs=None):
+    """Check arguments in argument_spec against those provided in module_paramaters
```
s/module_paramater/module_parameters/

samdoran replied:
One day I will be able to spell that right consistently.

samdoran committed:
Rework function behavior and documentation

I realized I missed the original intent of this method when moving it to a function. It
is meant to compared the parameter keys to legal inputs always, not compare
parameter keys to argument spec keys, even though the argument spec keys should
be a subset of legal inputs.

samdoran committed:
Add tests

samdoran committed:
Fix typo.

samdoran committed:
Set internal properties when handling suboptions

Shrews commented on lib/ansible/module_utils/basic.py:
```diff
@@ -1817,6 +1823,7 @@ def _handle_options(self, argument_spec=None, params=None, prefix=''):
 
                     options_legal_inputs = list(spec.keys()) + list(options_aliases.keys())
 
+                    self._set_internal_properties(spec, param)
```
Per our discussion, I think this will be fine to duplicate the call here. If it were contained within `_check_arguments` itself, we'd guarantee that we didn't accidentally call one without the other, and we could make the args to `_set_internal_properties` non-optional, but I also don't see us needing to add any more calls to `_check_arguments` either. So either way you prefer I suppose.

samdoran replied:
I like having them separate as separate calls. It keeps the methods focused rather than trying to do too many things.

samdoran committed:
Use mock in test

samdoran commented:
CI failures are due to an [openSUSE outage](https://twitter.com/openSUSE/status/1329078508971122696), unrelated to this PR.
